### DISCUSSION TITLE
useStore of a vanilla store sometimes doesn't react to updates

### DISCUSSION THREAD

**Discussion by elieven** [pmndrs/zustand#2467] [Bug report]
I'm making use of vanilla stores (multiple may be created and passed down to children via context) that connect to various remote updates. There seem to be some cases where `useStore` works and some where it doesn't while using these vanilla stores. 

I think this is most likely due to some misunderstanding I have about the workings but you never know.

I am not even entirely sure how to describe what the situation is where it doesn't work so forgive me - but I'll ask you to look at two minimal reproductions side by side.

Here is [the working version one](https://codesandbox.io/p/sandbox/zustand-vanilla-store-react-use-7qt22h?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cluopybuw0006356k4gddqhsa%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cluopybuw0002356ky0s6f36o%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cluopybuw0003356knb5jzuxb%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cluopybuw0005356k4jhm0d84%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B54.78188401632278%252C45.21811598367722%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cluopybuw0002356ky0s6f36o%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cluopybuv0001356k9ugj6q6l%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522cluopybuw0002356ky0s6f36o%2522%252C%2522activeTabId%2522%253A%2522cluopybuv0001356k9ugj6q6l%2522%257D%252C%2522cluopybuw0005356k4jhm0d84%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cluopybuw0004356klc618zfe%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cluopybuw0005356k4jhm0d84%2522%252C%2522activeTabId%2522%253A%2522cluopybuw0004356klc618zfe%2522%257D%252C%2522cluopybuw0003356knb5jzuxb%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cluopybuw0003356knb5jzuxb%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) - the value of the context is the store directly and in this case `useStore` works and the UI updates. And here is [the one that does not work](https://codesandbox.io/p/sandbox/zustand-vanilla-store-react-use-forked-vzrkdn?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cluopybuw0006356k4gddqhsa%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cluopybuw0002356ky0s6f36o%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cluopybuw0003356knb5jzuxb%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cluopybuw0005356k4jhm0d84%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B54.78188401632278%252C45.21811598367722%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cluopybuw0002356ky0s6f36o%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cluopybuv0001356k9ugj6q6l%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522cluopybuw0002356ky0s6f36o%2522%252C%2522activeTabId%2522%253A%2522cluopybuv0001356k9ugj6q6l%2522%257D%252C%2522cluopybuw0005356k4jhm0d84%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cluopybuw0004356klc618zfe%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cluopybuw0005356k4jhm0d84%2522%252C%2522activeTabId%2522%253A%2522cluopybuw0004356klc618zfe%2522%257D%252C%2522cluopybuw0003356knb5jzuxb%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cluopybuw0003356knb5jzuxb%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D). The only difference here is that the store is inside an object. 

Strangely enough sometimes it randomly starts working and updating when the codesandbox or local hot module replacement kicks in.

Any idea what might be the problem?

**Comment by dai-shi** [comment]
Can't see the codesandbox, but it sounds like you mutate the object, where you should treat it immutably, just guessing.
